cognito/cognito.go

```diff
@@ -0,0 +1,312 @@
+package cognito
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
+
+	"github.com/Ryanair/goaws"
+)
+
+var (
+	emailDeliveryMethod = cognitoidentityprovider.DeliveryMediumTypeEmail
+	smsDeliveryMethod   = cognitoidentityprovider.DeliveryMediumTypeSms
+
+	DeliveryMediumEmpty       = deliveryMedium([]*string{})
+	DeliveryMediumEmail       = deliveryMedium([]*string{&emailDeliveryMethod})
+	DeliveryMediumSms         = deliveryMedium([]*string{&smsDeliveryMethod})
+	DeliveryMediumEmailAndSms = deliveryMedium([]*string{&emailDeliveryMethod, &smsDeliveryMethod})
+)
+
+type deliveryMedium []*string
+
+type AuthenticationResult struct {
+	AccessToken  *string
+	ExpiresIn    *int64
+	IDToken      *string
+	RefreshToken *string
+	TokenType    *string
+}
+
+type SignInResult struct {
+	AuthenticationResult *AuthenticationResult
+	ChallengeName        *string
+	ChallengeParameters  map[string]*string
+	Session              *string
+}
+
+type CreateUserResult struct {
+	Attributes       map[string]string
+	Enabled          *bool
+	CreateDate       *time.Time
+	LastModifiedDate *time.Time
+	UserStatus       *string
+	Username         *string
+}
+
+type GetUserResult struct {
+	UserAttributes map[string]string
+	Username       *string
+}
+
+type Adapter struct {
+	poolID       string
+	clientID     string
+	clientSecret string
+	provider     *cognitoidentityprovider.CognitoIdentityProvider
+}
+
+func NewAdapter(cfg *goaws.Config, poolID, clientID, clientSecret string) *Adapter {
+
+	provider := cognitoidentityprovider.New(cfg.Provider)
+
+	return &Adapter{
+		poolID:       poolID,
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		provider:     provider,
+	}
+}
+
+func (ca *Adapter) ChangePassword(username, oldPassword, newPassword string) error {
+
+	adminAuthResponse, err := ca.SignIn(username, oldPassword)
+	if err != nil {
+		return wrapOpsErr(err, "error in cognito.Adapter while changing password")
+	}
+
+	switch *adminAuthResponse.ChallengeName {
+	case cognitoidentityprovider.ChallengeNameTypeNewPasswordRequired:
+		_, err := ca.respondToAuthChallenge(username, newPassword, adminAuthResponse.Session)
+		return err
```

Comment:
wrapErr to add some context?

Reply:
In functions exposed to clients I would use wrapOpsErr/wrapErr because they can return errors that represent some behaviour. From internal functions I would return wrapped errors with errors.Wrap

Reply:
updated